### ISSUE TITLE
Gammapy import fails on Binder

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#726]
Currently `import gammapy` fails in the IPython notebooks on Binder.

https://github.com/gammapy/gammapy-extra#binder

The cause is unknown, `import astropy` works.

I'll look into the issue and fix this later this week.

**Comment by adonath** [comment]
It seems like none of the dependencies specified in the `Dockerfile` works. Astropy probably only works, because it is accidentally part of the base anaconda installation.

**Comment by cdeil** [comment]
@adonath - I'm debugging the issue locally now. No need for you to also investigate the same thing.

**Comment by cdeil** [comment]
Look like it's a known recent issue with conda install on Binder:
see https://github.com/binder-project/binder/issues/80
and discussion on https://gitter.im/binder-project/binder

Probably we'll just have to wait a day or two ...

Note to self: put some command in the Dockerfile to generate the MPL font cache, so that this warning doesn't come up at the start of each IPython notebook:
```
/home/main/anaconda2/envs/python3/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
```

**Comment by adonath** [comment]
@cdeil Jeremy Freeman gave a hint at gitter that the gammapy-extra Binder works on Python 2. And indeed that's the issue: if you manually change to a Python 2 kernel it works. Probably the default kernel on Binder changed from Python 2 to Python 3 and as we don't install the dependencies for Python 3 in our Dockerfile it fails..

**Comment by cdeil** [comment]
I've worked on this issue a bit:
https://github.com/gammapy/gammapy-extra/blob/master/Dockerfile

I tried to create a Python 3 conda env, but didn't manage to activate it propertly (because Docker is running `/bin/sh` which doesn't have `source`, not `/bin/bash`, and there's no `sudo`.

Now I changed to a Python 3.5 Docker base image with Anaconda already present:
https://hub.docker.com/r/andrewosh/binder-python-3.5

All install seems to be OK now if I test with Docker locally.
But I haven't figured out yet how to run binder locally.

I'm going home now, will check the result of the new build on Binder tonight:
http://mybinder.org/status/gammapy/gammapy-extra

**Comment by cdeil** [comment]
The build seems to work, but binder is not starting up correctly, it currently shows "failed" without further info.
I've asked on Gitter for help (https://gitter.im/binder-project/binder?at=57ffb58fb68c1ca566b1b93a) but no response yet.

**Comment by cdeil** [comment]
@adonath - Thanks for giving this another try. I've assigned the issue to you.

Above my notes, in case they are useful. In short - I'd suggest you try going back to Python 2, what you set up first, unless you think you understand the startup config issue for Python 3 or find a working other example built on top of a Python 3 Docker image.

**Comment by adonath** [comment]
I changed the docker build to use Python 2.7 and the build succeeds (http://mybinder.org/status/gammapy/gammapy-extra) again, but launching binder still doesn't work for me. I noticed that other binder projects such as [ligo-binder](http://mybinder.org/repo/minrk/ligo-binderGW150914_tutorial.ipynb) or [scikit-learn-binder](http://mybinder.org/repo/glouppe/tutorial-scikit-learn) don't work either.

**Comment by cdeil** [comment]
@adonath - Thanks for trying!

Yeah, this is really frustrating, to have a binder that doesn't start and there's no error message.

I guess the quickest way would be to start with a working one (even if very primitive) and to add things line by line to the Dockerfile (or bisect) to see what introduces the issue.

Or we give up on this for now.
(I certainly don't have time today.)

**Comment by cdeil** [comment]
@adonath - Thanks for looking into this.

My 2 cents: yes, we could probably ask for a JupyterHub running at MPIK, but I'm not sure if it's a good investment of our and MPIK IT time. Setting it up is easy / quick, but then there's maintenance / security updates / monitoring: every such server on the internet is under attach and astronomers aren't very careful with passwords. So my recommendation would be to keep mentioning this in CTA, e.g. for the second data challenge next year. CTA observatory has money and IT personnel and probably enough motivation (i.e. eventually TB of high-level data, much more low-level data) to set up and maintain something like this.

**Comment by cdeil** [comment]
This was done by @Bultako  in https://github.com/gammapy/gammapy-extra/pull/91 .

The link to start it is now here and works for me also:
https://github.com/gammapy/gammapy-extra#binder

There's a few tutorials that access data that isn't in Binder. But those should be changed anyways to access data in a different way, I don't think we have to / should clone all extra repos with data into the Binder. We can discuss on a case by case basis.

@Bultako - the biggest remaining question is how users get to the Binder / notebooks.

We have this which is already or is getting more and more outdated:
https://github.com/gammapy/gammapy-extra/blob/master/index.ipynb
I think we should remove it completely, and instead maintain index information for the notebooks at
https://github.com/gammapy/gammapy-extra/blob/master/notebooks/notebooks.yaml
or maybe also at
https://github.com/gammapy/gammapy/blob/master/docs/notebooks.rst

I'm not sure what to do for Binder. Is an index notebook inside the binder useful or needed?
If yes, can we auto-generate it from `notebooks.yaml` or avoid having to manually maintain more than one "index" page for tutorials in some other way?

If we start using `notebooks.yaml` also for generating index files of available notebooks, we could add fields and info as needed, e.g. "index_on_binder: True" or something like that?

**Comment by Bultako** [comment]
@cdeil 
I would propose the following.

- **Entry points** to *Binder Gammapy*:
        - From the *introduction to notebooks* page http://docs.gammapy.org/en/latest/notebooks.html
        - From every text-fixed sphinx formatted notebook page, that will have a *binder badge*.
        - From the Readme of `gammapy-extra` GitHub repo.
- *Binder Gammapy* will start with the Jupyter browser inside the `notebooks` folder, just listing the *.ipynb files, so all the remaining **structure/folders above will be hidden**. **No need of an index.ipynb, I would remove it**. The Jupyter browser does not need it (on the contrary as index.html in web servers), it does not contain any code, it needs continuous updates...I do not see any pros.
- The `notebooks` folder of the `gammapy-extra` repo will contain **only the notebooks exposed** in the *Binder Gammapy*, that are the same as those present in the *last version* of the docs.
- Any other notebooks that we do not want to include because of work in progress or any other reason, we may put them in another different folder of `gammapy-extra` repo.
- The list of notebooks will be presented/described (as rich-text) **only in the *introduction to notebooks* page**. So an addition/removal of notebooks in `notebooks` folder of the `gammapy-extra` repo would need an update of this page in the docs in `gammapy` repo. Automation of this sync will not provide us with flexibility to describe them with free-text in the *introduction to notebooks* page.
- We may continue using `notebooks.yaml` for testing purposes.

**Comment by cdeil** [comment]
@Bultako - That sounds good to me!

My remaining questions / concerns:

> From every text-fixed sphinx formatted notebook page, that will have a binder badge.

Directly to the notebook, or to the notebook file browser? Pointing directly to the notebook would be nice, but is it possible with binder or does there have to be a single entry point? And if the user then would want to more to another notebook, would they have to start a new session or can they do it in one given binder session via the jupyter menu?

What about http://docs.gammapy.org/en/latest/notebooks.html#list-of-notebooks ? This is kinda duplicating the nicely grouped / introduced notebooks at http://docs.gammapy.org/en/latest/notebooks.html#notebooks and I think maybe having a second list of notebooks is confusing for users? Is it even technically possible to remove it, or will the lack of the toctree directive break everything?

I'm strongly in favour of getting rid of `gammapy-extra/index.ipynb`, but would suggest to replace it for a while with an empty page saying "Gammapy notebook index is now here: http://docs.gammapy.org/en/latest/notebooks.html 

Last point: I think I would prefer http://docs.gammapy.org/en/latest/notebooks.html to be called http://docs.gammapy.org/en/latest/tutorials.html - now we're using notebooks, but who knows what we have in 10 years, and even soon we might have some tutorials that aren't notebooks, so I think "tutorials" seems the more general term and would prefer to have that as the long-term URL. Again, I think this is a simple rename of `notebooks.rst` -> `tutorials.rst`

Finally: @Bultako - do you have more bandwidth to implement this in the coming ~ week or should I take a stab at it?

**Comment by Bultako** [comment]
@cdeil 
some good news below..

> Directly to the notebook, or to the notebook file browser? 

Pointing directly to the notebook

> And if the user then would want to more to another notebook, would they have to start a new session or can they do it in one given binder session via the jupyter menu?

They do it in one given binder session via the jupyter menu

> will the lack of the toctree directive break everything?

I'll try to fix the warnings with this
http://nbsphinx.readthedocs.io/en/0.2.17/orphan.html

> do you have more bandwidth to implement this in the coming ~ week or should I take a stab at it?

I may find sometime, I think it's not so much work, I hope..

p.s.
I seems we have kind of an issue with the last compilation in RTD :(

**Comment by cdeil** [comment]
It's not clear from http://readthedocs.org/projects/gammapy/builds/6303684/ what the issue is. Looks like it hangs. Such things happen quite frequently on ReadTheDocs. By now my strategy is to always wait for a build or two and day or two and most of the time the issue was a fluke and goes away by itself. And only if that doesn't happen, I "wipe" the build there and investigate the build log in detail and communicate to the RTD people about the issue via their Github issue tracker.

So in short: I'll take care of that, please just ignore issues on RTD for a while and be patient.

**Comment by cdeil** [comment]
I'm closing this old issue now. Basically it's done: we have working binder now and most of the things discussed here are implemented.

If there's further things to discuss, I'd suggest to open new issues.
